src/language-js/needs-parens.js

```diff
@@ -77,6 +77,17 @@ function needsParens(path, options) {
       return true;
     }
 
+    // `(let)[a] = 1`
+    if (
+      name === "object" &&
+      node.name === "let" &&
+      parent.type === "MemberExpression" &&
+      parent.computed &&
+      !parent.optional
+    ) {
+      return path.callParent(isAssignmentExpressionLeft);
```

Comment:
I don't think you should detect assignment expressions. The syntax is 

```
ExpressionStatement :
  [lookahead ∉ { {, function, async [no LineTerminator here] function, class, let [ }] Expression ;
```

So you should be looking at any expression statement that starts with `let [`, which _I hope_ the condition here sufficiently covers. Some more test cases:

[playground](https://deploy-preview-14000--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEAKANnGBKA2gJgF0BuAHSgENcBGQgAgGo6MsD6BeO-bMqEAGhAQADjACW0AM7JQFAE5yIAdwAK8hNJQV0SigE9pggEZyKYANZYAyhQC2cADJiocZADNtkuMdMXrws2cAc2QYOQBXbxA4WyM4ABN4hIcKKCDwiiC4ADEIOVsKGHE05BAKcJgIARAACxhbdAB1GrF4SQCwOCsNVrEAN1a9UrBJQxBnLzkYFVMggvdPKIArSQAPK2DMAEVwiHgF9C9BALlJ0qMKOPRq4TlnGEaxeJga5AAOAAZjxS9G02FSrc4JM+q5BABHXbwGYiTRlSQAWhcCQS1TkcEhYnRM0y8yQHkOUS8tjEoQiRM2cAAgkU7kYKnAVHA5E4XAcjiBJJSdntXPjFoIYJdHs9Xkh8ILTGJ0MEAMIQWx46KSACs1XCXgAKpdNASOX1IgBJKBJWBWMB3URUk1WGB6TDsuAAXydQA)

```js
(let)[2];
a[1] + (let[2] = 2);
```

IMO, both should have no change, but at least the first one needs fix.